pulsar-client/src/main/java/org/apache/pulsar/client/impl/UnAckedMessageRedeliveryTracker.java

```diff
@@ -0,0 +1,242 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+import io.netty.util.Timeout;
+import io.netty.util.TimerTask;
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.RedeliveryBackoff;
+import org.apache.pulsar.client.impl.conf.ConsumerConfigurationData;
+import org.apache.pulsar.common.util.collections.ConcurrentOpenHashSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UnAckedMessageRedeliveryTracker extends UnAckedMessageTracker {
+
+    private static final Logger log = LoggerFactory.getLogger(UnAckedMessageRedeliveryTracker.class);
+
+    protected final ConcurrentHashMap<UnackMessageIdWrapper, ConcurrentOpenHashSet<UnackMessageIdWrapper>>
+            redeliveryMessageIdPartitionMap;
+    protected final ArrayDeque<ConcurrentOpenHashSet<UnackMessageIdWrapper>> redeliveryTimePartitions;
+
+    protected final HashMap<MessageId, Long> ackTimeoutMessages;
+
+    private final RedeliveryBackoff ackTimeoutRedeliveryBackoff;
+
+    public UnAckedMessageRedeliveryTracker(PulsarClientImpl client, ConsumerBase<?> consumerBase,
+                                           ConsumerConfigurationData<?> conf) {
+        super(client, consumerBase, conf);
+        this.ackTimeoutRedeliveryBackoff = conf.getAckTimeoutRedeliveryBackoff();
+        this.ackTimeoutMessages = new HashMap<MessageId, Long>();
+        this.redeliveryMessageIdPartitionMap = new ConcurrentHashMap<>();
```

Comment:
Does this `redeliveryMessageIdPartitionMap` need to be ConcurrentHashMap ? 
It seems to be wrapped with `writeLock`

Reply:
can remove concurrent var.